### ISSUE TITLE
Inexact name of user defined types in AST

### ISSUE THREAD

**Issue by asinyagin** [ethereum/solidity#1364]
Hi,

If user defined type is deducted (`var s = S(1, 2);`), it's printed in AST without its contract name. So, in that case:
```
pragma solidity ^0.4.0;

contract Contract1 {
  enum E { E1, E2 }
}

contract Contract {
  enum E { E1, E2 }
  function test() {
    var e1 = Contract1.E.E1;
    var e3 = E.E1;
  }
}
```
the variables in AST will have the same type:
```
{
    "attributes": {
        "name": "e1",
        "type": "enum E"
    },
    "children": [],
    "id": 12144776,
    "name": "VariableDeclaration",
    "src": "133:6:1"
}
...
{
    "attributes": {
        "name": "e2",
        "type": "enum E"
    },
    "children": [],
    "id": 12145736,
    "name": "VariableDeclaration",
    "src": "162:6:1"
}
```

I think the type should be like `enum Contract1.E`.

```
root@4a541893579d:~/workspace/example-project/contracts# solc --version
solc, the solidity compiler commandline interface
Version: 0.4.5-develop.2016.11.9+commit.c82acfd3.Linux.g++
```

**Comment by chriseth** [comment]
In the source, this is called the "canonical name".

Feel free to grab that, anyone.

**Comment by asinyagin** [comment]
I'll send a PR in 10 mins

**Comment by axic** [comment]
Fixed by #1367.
